HW #2 (SOLID and GRASP)

cholombytko opened the pull request:
This pull request will contain changes that will be occured during refactoring code according to SOLID and GRASP principles.

cholombytko committed:
refactor: refactor mailing service using solid

This refactor contains creation of new interface which describes mailing
service using Dependency Inversion principle

cholombytko committed:
refactor: refactor rate service using solid

This commit contains refactoring of rate service using Dependency
Inversion principle

cholombytko committed:
refactor: refactor email service using solid

This commit contains refactoring of email service using Dependency
Inversion principle

cholombytko committed:
refactor: refactor several services using solid

This commit contains several updates, they occured by using SRP(Single
responsibility principle)

cholombytko committed:
refactor: refactor schedule service using solid

This commit contains refactoring of schedule service using Dependency
Inversion principle

cholombytko committed:
refactor: refactor subscription service using solid

This commit contains refactoring of subscription service using
Dependency Inversion principle

eluyema commented on src/email/services/mailing.service.ts:
```diff
@@ -1,11 +1,18 @@
+    private readonly emailService: IEmailService,
+  ) {}
 
   public async sendMail(payload: ISendMail): Promise<void> {
```
Maybe you can make this method private

cholombytko replied:
I think this method could be useful in the future, so I prefer to keep the public accessor

eluyema commented on src/email/services/mailing.service.ts:
```diff
@@ -19,4 +26,31 @@ export class MailingService {
+      }),
+    );
+
+    await Promise.all(mailPromises);
```
I think, you should consider using Promise.allSettled here, because the sending of each email does not depend on the others. For example, when using Promise.all, if any email throws an error, it can stop the sending of the other emails.

eluyema commented on src/email/services/mailing.service.ts:
```diff
@@ -19,4 +26,31 @@ export class MailingService {
+
+  public async sendMailsToSubscribers(): Promise<void> {
+    const { currencyCode, date, rate } =
+      await this.rateService.getExchangeRate();
```
I think it's better to move this responsibility to another service. The simplest way is to move getExchangeRate(); into public async scheduleEveryDayTask(): Promise<void>, and pass this data to the method through arguments

cholombytko replied:
Good catch, thanks!

cholombytko committed:
refactor: refactor sendmailstosubscribers method

This method contains refactor of SendMailsToSubscribers using
Promise.allSettled() instead Promise.all()

cholombytko committed:
refactor: move rate exchange responsibility

This commit contains refactoring of sendMailsToSubscribers method by
moving getExchangeRate to schedule service

cholombytko committed:
refactor: addon to previous commit

Larisa-Antsifrova commented on src/email/services/mailing.service.ts:
```diff
@@ -19,4 +27,30 @@ export class MailingService {
+    date: string,
+  ): string {
+    return `<p>Current rate ${currencyCode} to UAH - ${rate}. Date: ${date}</p>`;
+  }
```
At the moment, this service is less flexible than the others:
- Starting with clearly stated currencies
- Up to sending emails to active subscribers only

Larisa-Antsifrova commented on src/email/services/mailing.service.ts:
```diff
@@ -19,4 +27,30 @@ export class MailingService {
+  public async sendMailsToSubscribers(payload: IRate): Promise<void> {
+    const { currencyCode, date, rate } = payload;
+
+    const mailMessage = this.mailHtmlTemplate(currencyCode, rate, date);
```
Probably, the parameters for this method could also be wrapped in an object for consistency)
